### ISSUE TITLE
T056 - Add Graph-Aware ML Feature and Artifact Contracts

### ISSUE THREAD

**Issue by Scottlr** [Scottlr/grand-edge#60]
﻿

# T056 - Add Graph-Aware ML Feature and Artifact Contracts

## Objective
Extend the ML and model-runtime contracts to support graph-derived features, edge stability features, event features, graph rankers, and future GNN experiments while keeping MVP inference interpretable and Rust-owned.

## Scope
- Add graph feature schema fields to Python export and Rust model-runtime validation.
- Add graph model artifact metadata for graph version, graph feature version, relation corpus hash, and edge-observation window.
- Add GBM/ranker feature groups for own item features, neighbor features, sector features, conversion features, shock features, edge stability features, and event features.
- Add explicit runtime validation errors for unsupported graph neural network artifacts and missing graph provenance.
- Document that graph neural networks are deferred until enough labelled outcomes and graph metrics exist.

## Out of Scope
- Do not implement a graph neural network in MVP.
- Do not add Python to production inference.
- Do not require graph features for baseline recommendations.
- Do not train real production graph models without reviewed datasets.

## Blocked By

- #39 - T017 Add Offline Model Artifact and Adaptive Allocator Contracts
- #40 - T021 Create Python ML Research Workspace and Export Pipeline
- #41 - T022 Add Rust Model Runtime and Artifact Inference Bridge
- #48 - T049 Compute Graph-Derived Feature Vectors
- #50 - T051 Add Learned Edge Discovery and Stability Scoring
- #59 - T055 Add Graph Path Metrics and Outcome Evaluation

## Blocking

None

## Why It Is Blocked
Graph-aware model artifacts need base artifact contracts, Python export, Rust runtime validation, graph feature keys, learned-edge observations, and graph path metrics before they can be trustworthy.

## Related Code
- `crates/strategies/src/artifacts.rs` â€” artifact metadata from T017.
- `ml/src/grandedge_ml/artifact_schema.py` â€” Python artifact schema.
- `ml/src/grandedge_ml/features.py` â€” feature schema export.
- `crates/model_runtime/src/schema.rs` â€” Rust schema validation.
- `crates/features/src/graph.rs` â€” graph feature keys.
- `crates/metrics/src/graph.rs` â€” graph performance metrics.

## Implementation Touchpoints
- Read first: T017, T021, T022, T049, T051, T055, Python artifact schema, and Rust runtime schema validation.
- Edit first: `ml/src/grandedge_ml/artifact_schema.py`, then `features.py`, `export.py`, and `crates/model_runtime/src/schema.rs`.
- Symbols to search: `ModelArtifactMetadata`, `feature_schema_hash`, `FEATURE_SET_VERSION`, `GRAPH_FEATURE_SET_VERSION`, `GraphVersion`, `future_actionable_return`.
- Required structural changes: artifact metadata fields, feature schema groups, Rust validation error variants, fixture artifacts, model-card wording checks, and tests.
- Do not edit: live API, recommender scoring, graph UI, or production deployment.
- Verify in order: `uv run --project ml pytest`, `cargo test -p model_runtime graph_schema`, `cargo check --workspace`.
- Stop and ask: before adding GNN frameworks, PyTorch, remote feature stores, or mandatory ONNX graph models.

## Implementation Details
- Extend model card:

```json
{
  "strategy_id": "graph_ranker_v1",
  "model_version": "2026-06-16.1",
  "feature_set_version": "features_v1",
  "graph_feature_set_version": "graph_features_v1",
  "graph_version": "relations_2026_06_16_v3",
  "relation_corpus_hash": "sha256:...",
  "edge_observation_window_start": "2026-05-01T00:00:00Z",
  "edge_observation_window_end": "2026-06-01T00:00:00Z",
  "target": "future_actionable_return_6h",
  "known_limitations": [
    "Learned edges are predictive evidence, not causality.",
    "Observed volume is not true market depth."
  ]
}
```

- Feature groups:
Add a typed feature group enum in Rust and matching Python constants:

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum GraphFeatureGroup {
    OwnItemFeatures,
    ObservedExecutionProxyFeatures,
    NeighborReturnFeatures,
    SectorFeatures,
    ConversionFeatures,
    ShockFeatures,
    EdgeStabilityFeatures,
    EventFeatures,
    MissingDataFlags,
}
```

Feature group names:

```text
own_item_features
observed_execution_proxy_features
neighbor_return_features
sector_features
conversion_features
shock_features
edge_stability_features
event_features
missing_data_flags
```

- Required graph features:

```text
upstream_pressure_1h
downstream_pressure_1h
relative_value_residual
conversion_gap_pct
graph_adjusted_momentum_6h
link_disagreement_6h
strongest_graph_path_confidence
graph_neighbor_count
graph_missing_neighbor_data_count
edge_stability_score
event_exposure_score
```

- Add explicit graph metadata struct:

```rust
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GraphArtifactMetadata {
    pub graph_feature_set_version: String,
    pub graph_version: String,
    pub relation_corpus_hash: String,
    pub edge_observation_window_start: DateTime<Utc>,
    pub edge_observation_window_end: DateTime<Utc>,
    pub graph_feature_groups: Vec<GraphFeatureGroup>,
}
```

- Non-graph artifacts should store `graph: Option<GraphArtifactMetadata>` as `None`. Graph artifact kinds must require `Some(graph)`.

- Rust runtime validation must reject graph artifacts when:
  - `graph_feature_set_version` is missing for graph model kinds;
  - `graph_version` is missing;
  - relation corpus hash is missing;
  - feature schema lacks graph feature group metadata;
  - training/evaluation windows overlap incorrectly;
  - model card claims causal learned edges.
- Add validation error variants:

```rust
#[derive(Debug, thiserror::Error)]
pub enum ModelArtifactValidationError {
    #[error("graph artifact missing graph metadata")]
    MissingGraphMetadata,
    #[error("graph artifact missing relation corpus hash")]
    MissingRelationCorpusHash,
    #[error("graph feature group metadata is incomplete")]
    MissingGraphFeatureGroups,
    #[error("graph neural network artifacts are deferred")]
    GraphNeuralNetworkDeferred,
    #[error("learned edge wording must be predictive, not causal")]
    CausalLearnedEdgeClaim,
}
```

- Add `ModelArtifactKind::GraphRanker` and reserve `GraphNeuralNetwork` as disabled/deferred if useful:

```rust
#[serde(rename_all = "snake_case")]
pub enum ModelArtifactKind {
    GbdtRanker,
    GraphRanker,
    ContextualBandit,
    OnlineEnsemble,
    MetaLabel,
    GraphNeuralNetworkDeferred,
}
```

## Before / After

**Before:**
```text
Model artifacts know feature set version and schema hash.
Graph version and relation corpus provenance are not validated.
```

**After:**
```text
Graph-aware artifacts validate graph feature set, graph version, relation corpus hash, edge observation window, and graph limitation notes.
```

## Migration Notes
- Existing non-graph artifacts remain valid without graph metadata.
- Graph metadata is required only for graph artifact kinds.
- GNN artifacts must fail closed until a future task explicitly enables them.
- Adding graph metadata to the model card schema must be backward-compatible for non-graph artifact fixtures.

## Misinterpretations To Prevent
- Wrong interpretation: "GraphRanker means GNN."
  Correct rule: graph ranker can be a GBM/ranker using explicit graph features; GNN is deferred.
- Wrong interpretation: "Graph artifact can omit graph version because features are numeric."
  Correct rule: graph version and relation corpus hash are mandatory provenance.
- Wrong interpretation: "Python can serve graph models live."
  Correct rule: Python exports artifacts; Rust validates and serves them.
- Wrong interpretation: "A graph model card can use causal language if the metric is strong."
  Correct rule: learned-edge model cards must say predictive/non-causal unless the relation is mechanical/curated and source-backed.
- Wrong interpretation: "Graph feature names can drift between Python and Rust."
  Correct rule: Python feature export and Rust runtime validation share the same group names and required feature list through schemas/snapshots.

## Invariants
- Graph-aware artifacts include graph version and graph feature set version.
- Learned-edge features must be labelled predictive/non-causal in model cards.
- Graph features must not be mandatory for baseline strategies.
- GNN support is disabled/deferred until labelled outcomes and graph metrics justify it.
- Rust runtime remains the production inference boundary.
- Non-graph artifacts must continue validating without graph metadata.
- Graph artifacts must fail closed when graph provenance is missing or ambiguous.

## Acceptance Criteria
- [ ] Python artifact schema supports graph feature groups and graph model card fields.
- [ ] Rust runtime rejects graph artifacts missing graph version or relation corpus hash.
- [ ] Rust runtime rejects `GraphNeuralNetworkDeferred` artifacts with a dedicated unsupported/deferred error.
- [ ] Non-graph artifact fixtures remain valid without graph metadata.
- [ ] Fixture graph ranker artifact validates successfully.
- [ ] Fixture GNN artifact fails with deferred/unsupported error.
- [ ] Model card tests reject causal wording for learned edges.
- [ ] `uv run --project ml pytest` and `cargo test -p model_runtime graph_schema` pass.

## Testing Guidance
- Add Python `test_graph_artifact_requires_graph_version()`.
- Add Python `test_graph_feature_groups_include_path_features()`.
- Add Rust `graph_artifact_rejects_missing_relation_corpus_hash()`.
- Add Rust `graph_ranker_fixture_validates()`.
- Add Rust `non_graph_artifact_allows_missing_graph_metadata()`.
- Add Rust `gnn_artifact_is_deferred()`.
- Run `uv run --project ml pytest`.
- Run `cargo test -p model_runtime graph_schema`.
- Run `cargo check --workspace`.

## Agent Handoff Context
Graph-aware ML is powerful only after explicit graph features and graph metrics exist. Keep the first version interpretable: graph rankers over named features, not neural graph magic. GNNs can wait.

## Final Notes

- Recommended implementation order: after graph metrics and learned-edge stability exist.
- Unresolved questions: accepted graph artifact kinds and whether graph ranker artifacts live with other model artifacts.
- Risks: graph ML can obscure provenance if model cards skip graph version and relation hash.
- Areas that need human review before implementation: whether/when to enable GNN experiments.
